server/src/main/java/org/opensearch/action/support/replication/ReplicationOperation.java

```diff
@@ -226,13 +245,56 @@ private void performOnReplicas(
 
         final ShardRouting primaryRouting = primary.routingEntry();
 
-        for (final ShardRouting shard : replicationGroup.getReplicationTargets()) {
-            if (shard.isSameAllocation(primaryRouting) == false) {
-                performOnReplica(shard, replicaRequest, globalCheckpoint, maxSeqNoOfUpdatesOrDeletes, pendingReplicationActions);
+        for (final ReplicationAwareShardRouting shardRouting : replicationGroup.getReplicationTargets()) {
+            new PerformOnReplicaProxyFactory().create(
+                primaryRouting,
+                shardRouting,
+                replicaRequest,
+                globalCheckpoint,
+                maxSeqNoOfUpdatesOrDeletes,
+                replicationGroup,
+                pendingReplicationActions
```

Comment:
Can we inject this proxy in the ctor for `ReplicationOperation ` itself?

Reply:
@Bukhtawar - this would be difficult since the PerformOnReplicaProxy implementation changes per ShardRouting depending on TransportReplicationMode (need for replication) i.e. the replication strategy is not universal to all ShardRoutings per replication request.

ReplicationOperation would lead to a universal strategy for the request which is above the granularity at which we need PerformOnReplicaProxy.